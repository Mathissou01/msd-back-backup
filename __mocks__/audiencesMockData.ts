import { GetAudiencesByContractIdDocument } from "../src/graphql/codegen/generated-types";

export const getNoAudienceMockedData = [
  {
    request: {
      query: GetAudiencesByContractIdDocument,
      variables: {
        filters: {
          contract: {
            id: {
              eq: undefined,
            },
          },
          isActive: {
            eq: true,
          },
        },
      },
    },
    result: {
      data: {
        audiences: {
          data: [],
        },
      },
    },
  },
];

export const getSingleAudienceMockedData = [
  {
    request: {
      query: GetAudiencesByContractIdDocument,
      variables: {
        filters: {
          contract: {
            id: {
              eq: undefined,
            },
          },
          isActive: {
            eq: true,
          },
        },
      },
    },
    result: {
      data: {
        audiences: {
          data: [
            {
              id: 1,
              attributes: {
                isActive: true,
                type: "Particuliers",
                __typename: "Audience",
              },
            },
          ],
        },
      },
    },
  },
];

export const getTwoAudiencesMockedData = [
  {
    request: {
      query: GetAudiencesByContractIdDocument,
      variables: {
        filters: {
          contract: {
            id: {
              eq: undefined,
            },
          },
          isActive: {
            eq: true,
          },
        },
      },
    },
    result: {
      data: {
        audiences: {
          data: [
            {
              id: 1,
              attributes: {
                isActive: true,
                type: "Particuliers",
                __typename: "Audience",
              },
            },
            {
              id: 2,
              attributes: {
                isActive: true,
                type: "Collectifs",
                __typename: "Audience",
              },
            },
          ],
        },
      },
    },
  },
];

export const getAllAudiencesMockedData = [
  {
    request: {
      query: GetAudiencesByContractIdDocument,
      variables: {
        filters: {
          contract: {
            id: {
              eq: undefined,
            },
          },
          isActive: {
            eq: true,
          },
        },
      },
    },
    result: {
      data: {
        audiences: {
          data: [
            {
              id: 1,
              attributes: {
                isActive: true,
                type: "Particuliers",
                __typename: "Audience",
              },
            },
            {
              id: 2,
              attributes: {
                isActive: true,
                type: "Collectifs",
                __typename: "Audience",
              },
            },
            {
              id: 3,
              attributes: {
                isActive: true,
                type: "Professionnels",
                __typename: "Audience",
              },
            },
          ],
        },
      },
    },
  },
];
