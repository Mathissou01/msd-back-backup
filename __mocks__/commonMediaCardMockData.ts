import { GetFilesPaginationByFolderIdDocument } from "../src/graphql/codegen/generated-types";

export const defaultMockData = [
  {
    request: {
      query: GetFilesPaginationByFolderIdDocument,
      filters: {
        folder: {
          id: {
            eq: 2,
          },
        },
      },
      pagination: {
        page: 1,
        pageSize: 10,
      },
      sort: "createdAt:desc",
    },
    result: {
      data: {
        uploadFiles: {
          data: [
            {
              id: "124",
              attributes: {
                name: "La Bruja.jpeg",
                mime: "image/jpeg",
                size: 124.35,
                width: 1280,
                height: 853,
                createdAt: "2023-01-26T10:34:26.058Z",
                url: "https://stomsdmediadev.blob.core.windows.net/msd-media-dev/assets-local-sdial/La%20Bruja.jpeg",
                ext: ".jpeg",
              },
            },
            {
              id: "122",
              attributes: {
                name: "Banana Mama Rooftop & Kitchen 5.jpeg",
                mime: "image/jpeg",
                size: 68.84,
                width: 960,
                height: 960,
                createdAt: "2023-01-26T10:30:22.701Z",
                url: "https://stomsdmediadev.blob.core.windows.net/msd-media-dev/assets-local-sdial/Banana%20Mama%20Rooftop%20%26%20Kitchen%205.jpeg",
                ext: ".jpeg",
              },
            },
            {
              id: "121",
              attributes: {
                name: "Sichuan Hot Pot.jpeg",
                mime: "image/jpeg",
                size: 2233.13,
                width: 6962,
                height: 4615,
                createdAt: "2023-01-18T17:39:30.518Z",
                url: "https://stomsdmediadev.blob.core.windows.net/msd-media-dev/assets-local-sdial/Sichuan%20Hot%20Pot.jpeg",
                ext: ".jpeg",
              },
            },
            {
              id: "120",
              attributes: {
                name: "mahbod-akhzami-kg40JrEsCvY-unsplash.jpg",
                mime: "image/jpeg",
                size: 1788.48,
                width: 6000,
                height: 4000,
                createdAt: "2023-01-18T17:39:29.900Z",
                url: "https://stomsdmediadev.blob.core.windows.net/msd-media-dev/assets-local-sdial/mahbod-akhzami-kg40JrEsCvY-unsplash.jpg",
                ext: ".jpg",
              },
            },
            {
              id: "119",
              attributes: {
                name: "Paella in Valencia.jpeg",
                mime: "image/jpeg",
                size: 1026.93,
                width: 3851,
                height: 2528,
                createdAt: "2023-01-18T17:39:29.720Z",
                url: "https://stomsdmediadev.blob.core.windows.net/msd-media-dev/assets-local-sdial/Paella%20in%20Valencia.jpeg",
                ext: ".jpeg",
              },
            },
            {
              id: "118",
              attributes: {
                name: "Rusconi's American Kitchen 3.jpeg",
                mime: "image/jpeg",
                size: 209.81,
                width: 1999,
                height: 1333,
                createdAt: "2023-01-18T17:39:29.502Z",
                url: "https://stomsdmediadev.blob.core.windows.net/msd-media-dev/assets-local-sdial/Rusconi%27s%20American%20Kitchen%203.jpeg",
                ext: ".jpeg",
              },
            },
            {
              id: "117",
              attributes: {
                name: "Quince 5.jpeg",
                mime: "image/jpeg",
                size: 197.32,
                width: 2000,
                height: 1500,
                createdAt: "2023-01-18T17:39:29.310Z",
                url: "https://stomsdmediadev.blob.core.windows.net/msd-media-dev/assets-local-sdial/Quince%205.jpeg",
                ext: ".jpeg",
              },
            },
            {
              id: "116",
              attributes: {
                name: "Rusconi's American Kitchen 2.jpeg",
                mime: "image/jpeg",
                size: 72.49,
                width: 1000,
                height: 1500,
                createdAt: "2023-01-18T17:39:29.221Z",
                url: "https://stomsdmediadev.blob.core.windows.net/msd-media-dev/assets-local-sdial/Rusconi%27s%20American%20Kitchen%202.jpeg",
                ext: ".jpeg",
              },
            },
            {
              id: "115",
              attributes: {
                name: "Pur' - Jean-Fran√ßois Rouquette 3.jpeg",
                mime: "image/jpeg",
                size: 63.2,
                width: 1280,
                height: 853,
                createdAt: "2023-01-18T17:39:29.039Z",
                url: "https://stomsdmediadev.blob.core.windows.net/msd-media-dev/assets-local-sdial/Pur%27%20-%20Jean-Fran%C3%A7ois%20Rouquette%203.jpeg",
                ext: ".jpeg",
              },
            },
            {
              id: "114",
              attributes: {
                name: "muneeb-muhamed-kPXT9HA4LOk-unsplash.jpg",
                mime: "image/jpeg",
                size: 516.64,
                width: 3456,
                height: 4608,
                createdAt: "2023-01-18T17:39:28.922Z",
                url: "https://stomsdmediadev.blob.core.windows.net/msd-media-dev/assets-local-sdial/muneeb-muhamed-kPXT9HA4LOk-unsplash.jpg",
                ext: ".jpg",
              },
            },
          ],
          meta: {
            pagination: {
              total: 122,
              pageSize: 10,
              pageCount: 13,
              page: 1,
            },
          },
        },
      },
    },
  },
];
