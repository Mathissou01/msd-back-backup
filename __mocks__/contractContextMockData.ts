import {
  Enum_Contract_Clienttype,
  Enum_Contract_Contractstatus,
} from "../src/graphql/codegen/generated-types";

export const mockData = {
  id: "1",
  attributes: {
    clientName: "Grand test",
    clientType: "city" as Enum_Contract_Clienttype,
    contractStatus: "Initialisation" as Enum_Contract_Contractstatus,
    siret: 12345678910123,
    clear: 123456789,
    ccap: 123456789,
    isNonExclusive: false,
    isRVFrance: true,
    pathId: 2,
    logo: {
      data: null,
    },
    channelType: {
      data: {
        id: "1",
        attributes: {
          hasWebApp: false,
          hasWebSite: false,
        },
      },
    },
    clientContact: {
      data: {
        id: "1",
        attributes: {
          firstName: "test",
          lastName: "test",
          email: "test.test@mail.com",
          phoneNumber: "07 81 21 56 81",
        },
      },
    },
    contractCustomization: {
      data: {
        id: "1",
        attributes: {
          primaryColor: "#9BCD41",
          secondaryColor: "#FFC229",
          textContrast: "#030F40",
        },
      },
    },
    editorialService: {
      data: {
        id: "1",
        attributes: {
          eventSubService: {
            data: {
              id: "1",
              attributes: {
                name: "Evénement ",
                isActivated: true,
              },
            },
          },
          freeContentSubServices: {
            data: [
              {
                id: "1",
                attributes: {
                  name: "Editorial 1",
                  isActivated: true,
                },
              },
              {
                id: "2",
                attributes: {
                  name: "Editorial 2",
                  isActivated: true,
                },
              },
            ],
          },
          newsSubService: {
            data: {
              id: "1",
              attributes: {
                name: "Actualités",
                isActivated: true,
              },
            },
          },
          quizSubService: {
            data: {
              id: "1",
              attributes: {
                name: "Quiz",
                isActivated: true,
              },
            },
          },
          tipSubService: {
            data: {
              id: "1",
              attributes: {
                name: "Astuces",
                isActivated: true,
              },
            },
          },
        },
      },
    },
    recyclingGuideService: {
      data: {
        id: "1",
        attributes: {
          name: "Guide du tri",
          isActivated: false,
          memoName: "Memo Tri",
        },
      },
    },
    pickUpDayService: {
      data: {
        id: "1",
        attributes: {
          name: "Jour de collecte",
          isActivated: false,
        },
      },
    },
    dropOffMapService: {
      data: {
        id: "1",
        attributes: {
          name: "Points d'apports et carte",
          isActivated: false,
        },
      },
    },
    requestService: {
      data: {
        id: "1",
        attributes: {
          name: "Demande",
          isActivated: false,
        },
      },
    },
  },
};
