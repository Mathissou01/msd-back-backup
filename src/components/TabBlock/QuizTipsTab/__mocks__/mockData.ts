import { GetQuizAndTipsBlockDocument } from "../../../../graphql/codegen/generated-types";

export const defaultMockData = [
  {
    request: {
      query: GetQuizAndTipsBlockDocument,
      variables: {
        contractId: "1",
      },
    },
    result: {
      data: {
        contractCustomizations: {
          data: [
            {
              attributes: {
                homepage: {
                  data: {
                    attributes: {
                      quizAndTipsBlock: {
                        data: {
                          id: "1",
                          attributes: {
                            title: "Titre modifi√©",
                            displayBlock: false,
                            displayQuiz: false,
                            quiz: {
                              data: {
                                id: "2",
                                attributes: {
                                  title: "Nom du quiz",
                                  publishedAt: "2022-11-20T02:34:14.535Z",
                                },
                              },
                            },
                            displayTips: false,
                            tips: {
                              data: [],
                            },
                          },
                        },
                      },
                    },
                  },
                },
              },
            },
          ],
        },
        services: {
          data: [
            {
              attributes: {
                serviceInstance: [
                  {
                    __typename: "ComponentMsdEditorial",
                    id: "4",
                    editorialServices: {
                      data: [
                        {
                          attributes: {
                            subServiceInstance: [
                              {
                                __typename: "ComponentEditoQuizzesSubService",
                                isActivated: true,
                                quizzes: {
                                  data: [
                                    {
                                      id: "2",
                                      attributes: {
                                        title: "Nom du quiz",
                                        publishedAt: "2022-11-20T02:34:14.535Z",
                                      },
                                    },
                                    {
                                      id: "1",
                                      attributes: {
                                        title: "truc",
                                        publishedAt: "2022-11-18T10:16:08.936Z",
                                      },
                                    },
                                  ],
                                },
                              },
                            ],
                          },
                        },
                        {
                          attributes: {
                            subServiceInstance: [
                              {
                                __typename: "ComponentEditoTipsSubService",
                                isActivated: true,
                                tips: {
                                  data: [
                                    {
                                      id: "5",
                                      attributes: {
                                        title: "nah",
                                        publishedAt: "2022-11-19T00:14:37.403Z",
                                      },
                                    },
                                    {
                                      id: "2",
                                      attributes: {
                                        title: "Titre de l'astuce 1",
                                        publishedAt: "2022-11-19T00:14:08.232Z",
                                      },
                                    },
                                    {
                                      id: "1",
                                      attributes: {
                                        title: "Titre de l'astuce 2",
                                        publishedAt: "2022-11-18T10:17:11.158Z",
                                      },
                                    },
                                    {
                                      id: "3",
                                      attributes: {
                                        title: "Titre de l'astuce 3",
                                        publishedAt: "2022-11-19T00:14:20.490Z",
                                      },
                                    },
                                    {
                                      id: "4",
                                      attributes: {
                                        title: "Titre de l'astuce 4",
                                        publishedAt: "2022-11-19T00:14:30.273Z",
                                      },
                                    },
                                  ],
                                },
                              },
                            ],
                          },
                        },
                      ],
                    },
                  },
                ],
              },
            },
          ],
        },
      },
    },
  },
];

export const errorMockData = [
  {
    request: {
      query: GetQuizAndTipsBlockDocument,
      variables: {
        contractId: "1",
      },
    },
    error: new Error("An error occured"),
  },
];
