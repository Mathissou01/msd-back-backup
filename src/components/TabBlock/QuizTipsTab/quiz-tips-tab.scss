.c-QuizTipsTab {
  &__Title {
    margin-bottom: 18px;
    font-size: 28px;
  }

  &__Form {
    display: flex;
    flex-direction: column;
    gap: 24px;
    width: 100%;
  }

  &__Group {
    display: flex;
    flex-direction: column;
    gap: 10px;
  }

  &__Buttons {
    display: flex;
    flex-direction: row;
    gap: 16px;
  }
}
