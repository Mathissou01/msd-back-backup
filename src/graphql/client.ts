import {
  ApolloClient,
  InMemoryCache,
  NormalizedCacheObject,
} from "@apollo/client";

let client: ApolloClient<NormalizedCacheObject>;

if (process.env.NEXT_PUBLIC_MOCK === "true") {
  /* With .env variable NEXT_PUBLIC_MOCK=true, return mock graphql client with fake response data */
  client = (await import("./mockedClient")).default(1000);
} else {
  client = await new ApolloClient({
    uri: `${process.env.NEXT_PUBLIC_API_URL}/graphql`,
    cache: new InMemoryCache({
      typePolicies: {
        UploadFolder: {
          merge: true,
        },
        UploadFile: {
          merge: true,
        },
        Contract: {
          merge: true,
        },
        ContractCustomization: {
          merge: true,
        },
        New: {
          merge: true,
        },
        Tip: {
          merge: true,
        },
        RecyclingGuideService: {
          merge: true,
        },
        WasteForm: {
          merge: true,
        },
        WasteFamily: {
          merge: true,
        },
        Flow: {
          merge: true,
        },
        FlowColor: {
          merge: true,
        },
        DropOffMap: {
          merge: true,
        },
        Request: {
          merge: true,
        },
        RequestAggregate: {
          merge: true,
        },
      },
    }),
  });
}

export default client;
