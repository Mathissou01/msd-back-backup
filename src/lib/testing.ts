import { prettyDOM } from "@testing-library/react";

export function removeUnstableHtmlPropertiesForTesting(
  htmlElement: HTMLElement,
  attributeName: string,
) {
  const domHTML = prettyDOM(htmlElement, Infinity).toString();
  if (!domHTML) return undefined;
  const regex = new RegExp(`${attributeName}(.*)"(.*)"`, "g");
  return domHTML.replaceAll(regex, "");
}
