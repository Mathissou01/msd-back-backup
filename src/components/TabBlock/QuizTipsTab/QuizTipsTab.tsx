import { format, parseJSON } from "date-fns";
import { FormProvider, useForm } from "react-hook-form";
import { FieldValues } from "react-hook-form/dist/types/fields";
import React, {
  ReactNode,
  useCallback,
  useEffect,
  useRef,
  useState,
} from "react";
import {
  QuizEntity,
  TipEntity,
  GetQuizAndTipsBlockByContractIdDocument,
  useGetQuizAndTipsBlockByContractIdQuery,
  useUpdateQuizAndTipsBlockByIdMutation,
} from "../../../graphql/codegen/generated-types";
import { FocusFirstElement, removeNulls } from "../../../lib/utilities";
import { extractQuizAndTipsBlock } from "../../../lib/graphql-data";
import CommonButton from "../../Common/CommonButton/CommonButton";
import FormInput from "../../Form/FormInput/FormInput";
import FormCheckbox from "../../Form/FormCheckbox/FormCheckbox";
import FormModalInput from "../../Form/FormModalInput/FormModalInput";
import FormSelect from "../../Form/FormSelect/FormSelect";
import FormMultiselect from "../../Form/FormMultiselect/FormMultiselect";
import CommonSpinner from "../../Common/CommonSpinner/CommonSpinner";
import "./quiz-tips-tab.scss";

interface IQuizAndTipBlock {
  id: string;
  titleContent: string;
  displayBlock: boolean;
  displayQuiz: boolean;
  quiz?: QuizEntity | null;
  displayTips: boolean;
  tips?: Array<TipEntity> | null;
}

export default function QuizTipsTab() {
  /* Static Data */
  const formLabels = {
    title: "Quiz & Astuces",
    displayBlock: "Afficher ce bloc",
    titleContent: "Titre du bloc",
    displayQuiz: "Afficher un quiz",
    quiz: "Quiz à afficher",
    quizButton: "Choisir le quiz",
    quizModal: "Sélection du quiz",
    quizModalType: "Quiz",
    displayTips: "Afficher des astuces",
    tips: "Astuces à afficher",
    tipsButton: "Sélectionner les astuces à afficher",
    tipsModal: "Sélection des astuces",
    tipsModalType: "Astuce",
  };
  const submitButtonLabel = "Enregistrer les modifications";
  const cancelButtonLabel = "Annuler les modifications";

  /* Methods */
  function quizDisplayTransformFunction(
    quiz: QuizEntity | undefined,
  ): ReactNode {
    return <p>{quiz?.attributes?.title ?? ""}</p>;
  }

  function quizSelectDisplayTransformFunction(quiz: QuizEntity): string {
    return quiz.attributes?.title ?? "";
  }

  function onQuizModalSubmit(submitData: {
    [key: string]: QuizEntity | undefined;
  }) {
    const quiz = Object.values(submitData)[0];
    setValue("quiz", quiz, { shouldDirty: true });
  }

  function tipsDisplayTransformFunction(
    tips: Array<TipEntity | undefined>,
  ): ReactNode {
    return tips?.map((tip: TipEntity | undefined, index) => {
      if (tip && tip.id) {
        return (
          <p key={tip.id + index}>
            {`${tip.attributes?.title} - ${format(
              parseJSON(tip.attributes?.publishedAt),
              "dd/mm/yyyy",
            )}` ?? ""}
          </p>
        );
      }
    });
  }

  function tipSelectDisplayTransformFunction(tip: TipEntity): string {
    return (
      `${tip.attributes?.title} - ${format(
        parseJSON(tip.attributes?.publishedAt),
        "dd/mm/yyyy",
      )}` ?? ""
    );
  }

  function onTipsModalSubmit(submitData: {
    [key: string]: Array<TipEntity | undefined>;
  }) {
    const tips = Object.values(submitData)?.filter(removeNulls);
    setValue("tips", tips, { shouldDirty: true });
  }

  async function onSubmitValid(submitData: FieldValues) {
    if (submitData["id"]) {
      const variables = {
        quizAndTipsBlockId: submitData["id"],
        data: {
          title: submitData["titleContent"],
          displayBlock: submitData["displayBlock"],
          displayQuiz: submitData["displayQuiz"],
          quiz: submitData["quiz"]?.id ?? null,
          displayTips: submitData["displayTips"],
          tips:
            submitData["tips"]?.map((tip: TipEntity) => tip?.id ?? null) ??
            null,
        },
      };
      await updateQuizAndTipsBlock({
        variables,
        refetchQueries: [
          {
            query: GetQuizAndTipsBlockByContractIdDocument,
            variables: { contractId },
          },
          "GetQuizAndTipsBlock",
        ],
      });
      return new Promise<void>((resolve) => {
        setTimeout(() => resolve(), 1000);
      });
    }
  }

  function onCancel() {
    form.reset();
  }

  /* API Data */
  const contractId = "1"; // TODO: Put Contract data (ID) in STORE, maybe have hook to automatically insert ID variable in gql requests
  const { loading, error, data } = useGetQuizAndTipsBlockByContractIdQuery({
    variables: { contractId },
  });
  const [
    updateQuizAndTipsBlock,
    { loading: mutationLoading, error: mutationError },
  ] = useUpdateQuizAndTipsBlockByIdMutation();

  /* Local Data */
  const [isShowingSpinner, setIsShowingSpinner] = useState(false);
  const [quizTipsData, setQuizTipsData] = useState<IQuizAndTipBlock>();
  const [quizzesData, setQuizzesData] = useState<Array<QuizEntity>>([]);
  const [tipsData, setTipsData] = useState<Array<TipEntity>>([]);
  const formValidationMode = "onChange";
  const form = useForm({
    mode: formValidationMode,
  });
  const { handleSubmit, setValue, watch, formState } = form;
  const { isDirty, isSubmitting } = formState;

  useEffect(() => {
    if (data && data.contractCustomizations?.data && data.services?.data) {
      const { quizAndTipsBlock, quizzes, tips } = extractQuizAndTipsBlock(data);
      if (quizAndTipsBlock?.id && quizAndTipsBlock?.attributes) {
        const mappedData: IQuizAndTipBlock = {
          id: quizAndTipsBlock.id,
          titleContent:
            quizAndTipsBlock.attributes.titleContent ?? "Quiz & Astuces",
          displayBlock: quizAndTipsBlock.attributes.displayBlock,
          displayQuiz: quizAndTipsBlock.attributes.displayQuiz,
          quiz: quizAndTipsBlock.attributes.quiz?.data,
          displayTips: quizAndTipsBlock.attributes.displayTips,
          tips: quizAndTipsBlock?.attributes.tips?.data,
        };
        setQuizTipsData(mappedData);
        form.reset(mappedData);
      }
      if (quizzes && quizzes?.length > 0) {
        setQuizzesData(quizzes);
      }
      if (tips && tips?.length > 0) {
        setTipsData(tips);
      }
    }
  }, [form, data]);

  const spinnerTimerRef = useRef<NodeJS.Timeout>();
  useEffect(() => {
    if (isSubmitting) {
      if (!isShowingSpinner) {
        spinnerTimerRef.current = setTimeout(() => {
          setIsShowingSpinner(true);
        }, 200);
      }
    } else {
      clearTimeout(spinnerTimerRef.current);
      setIsShowingSpinner(false);
    }
  }, [isShowingSpinner, isSubmitting]);

  const focusRef = useCallback((node: HTMLFormElement) => {
    FocusFirstElement(node);
  }, []);

  {
    // TODO: layout shift, handle error redirect,
  }
  if (loading) return <CommonSpinner />;
  if (error) return <span>{error?.message}</span>;
  if (mutationError) return <span>{mutationError?.message}</span>;

  return (
    <div className="c-QuizTipsTab">
      {isShowingSpinner && <CommonSpinner isCover={true} />}
      <h2 className="c-QuizTipsTab__Title">{formLabels.title}</h2>
      <FormProvider {...form}>
        <form
          className="c-QuizTipsTab__Form"
          onSubmit={handleSubmit(onSubmitValid)}
          ref={focusRef}
        >
          <div className="c-QuizTipsTab__Group">
            <FormCheckbox name="displayBlock" label={formLabels.displayBlock} />
            <FormInput
              type="text"
              name="titleContent"
              label={formLabels.titleContent}
              isRequired={true}
              isDisabled={mutationLoading}
              defaultValue={quizTipsData?.titleContent}
            />
          </div>
          <div className="c-QuizTipsTab__Group">
            <FormCheckbox name="displayQuiz" label={formLabels.displayQuiz} />
            <FormModalInput<QuizEntity>
              name="quiz"
              label={formLabels.quiz}
              displayTransform={quizDisplayTransformFunction}
              buttonLabel={formLabels.quizButton}
              modalTitle={formLabels.quizModal}
              onSubmit={onQuizModalSubmit}
              formValidationMode={formValidationMode}
              isDisabled={mutationLoading}
            >
              <FormSelect<QuizEntity>
                name="quizSelect"
                label={formLabels.quizModalType}
                displayTransform={quizSelectDisplayTransformFunction}
                options={quizzesData}
                optionKey={"id"}
                defaultValue={watch("quiz")}
              />
            </FormModalInput>
          </div>
          <div className="c-QuizTipsTab__Group">
            <FormCheckbox name="displayTips" label={formLabels.displayTips} />
            <FormModalInput<Array<TipEntity>>
              name="tips"
              label={formLabels.tips}
              displayTransform={tipsDisplayTransformFunction}
              buttonLabel={formLabels.tipsButton}
              modalTitle={formLabels.tipsModal}
              onSubmit={onTipsModalSubmit}
              formValidationMode={formValidationMode}
              isDisabled={mutationLoading}
            >
              <FormMultiselect<TipEntity>
                name="tipsSelect"
                label={formLabels.tipsModalType}
                displayTransform={tipSelectDisplayTransformFunction}
                options={tipsData}
                selectAmount={3}
                optionKey={"id"}
                defaultValues={watch("tips")}
              />
            </FormModalInput>
          </div>
          <div className="c-QuizTipsTab__Buttons">
            <CommonButton
              type="submit"
              label={submitButtonLabel}
              style="primary"
              isDisabled={!isDirty}
            />
            <CommonButton
              type="button"
              label={cancelButtonLabel}
              onClick={onCancel}
              isDisabled={!isDirty}
            />
          </div>
        </form>
      </FormProvider>
    </div>
  );
}
