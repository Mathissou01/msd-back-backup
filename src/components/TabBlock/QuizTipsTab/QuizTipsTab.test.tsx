import { render, screen } from "@testing-library/react";
import { MockedProvider } from "@apollo/client/testing";
import React from "react";
import { defaultMockData } from "./__mocks__/mockData";
import QuizTipsTab from "./QuizTipsTab";

it("renders loading and loaded state", async () => {
  const { container } = render(
    <MockedProvider mocks={defaultMockData} addTypename={false}>
      <QuizTipsTab />
    </MockedProvider>,
  );

  expect(await screen.findByTestId("common-spinner")).toBeInTheDocument();
  expect(await container).toMatchSnapshot();
  expect(await screen.findByText("Quiz et Astuces")).toBeInTheDocument();
  expect(await container).toMatchSnapshot();
});

// it("should show error UI", async () => {
//   const { container } = render(
//     <MockedProvider mocks={errorMockData} addTypename={false}>
//       <QuizTipsTab />
//     </MockedProvider>,
//   );
//
//   expect(await screen.findByText("Loading...")).toBeInTheDocument();
//   expect(await screen.findByText("An error occurred")).toBeInTheDocument();
//   expect(await container).toMatchSnapshot();
// });
