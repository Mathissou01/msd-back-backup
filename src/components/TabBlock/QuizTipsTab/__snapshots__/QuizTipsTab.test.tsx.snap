// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders loading and loaded state 1`] = `
<div>
  <div
    class="c-CommonSpinner__Wrapper "
  >
    <div
      class="c-CommonSpinner"
      data-testid="common-spinner"
    >
      <div />
      <div />
      <div />
      <div />
    </div>
  </div>
</div>
`;

exports[`renders loading and loaded state 2`] = `
<div>
  <div
    class="c-QuizTipsTab"
  >
    <h2
      class="c-QuizTipsTab__Title"
    >
      Quiz & Astuces
    </h2>
    <form
      class="c-QuizTipsTab__Form"
    >
      <div
        class="c-QuizTipsTab__Group"
      >
        <div
          class="c-FormCheckbox"
        >
          <input
            class="c-FormCheckbox__Input "
            data-testid="form-checkbox"
            id="displayBlock"
            name="displayBlock"
            type="checkbox"
          />
          <label
            class="c-FormLabel"
            for="displayBlock"
          >
            <div
              class="c-FormLabel__Label"
            >
              <span>
                Afficher ce bloc 
              </span>
            </div>
          </label>
        </div>
        <div
          class="c-FormInput"
        >
          <label
            class="c-FormLabel"
            for="blockTitle"
          >
            <div
              class="c-FormLabel__Label"
            >
              <span>
                Titre du bloc *
              </span>
            </div>
          </label>
          <input
            aria-errormessage="blockTitle_error"
            aria-invalid="false"
            class="c-FormInput__Input "
            data-testid="form-input"
            id="blockTitle"
            name="blockTitle"
            type="text"
            value="Titre modifié"
          />
        </div>
      </div>
      <div
        class="c-QuizTipsTab__Group"
      >
        <div
          class="c-FormCheckbox"
        >
          <input
            class="c-FormCheckbox__Input "
            data-testid="form-checkbox"
            id="displayQuiz"
            name="displayQuiz"
            type="checkbox"
          />
          <label
            class="c-FormLabel"
            for="displayQuiz"
          >
            <div
              class="c-FormLabel__Label"
            >
              <span>
                Afficher un quiz 
              </span>
            </div>
          </label>
        </div>
        <div
          class="c-FormModalInput"
          data-testid="form-modal-input"
        >
          <label
            class="c-FormLabel"
            for="quiz_button"
          >
            <div
              class="c-FormLabel__Label"
            >
              <span>
                Quiz à afficher 
              </span>
            </div>
          </label>
          <input
            class="c-FormModalInput__Input"
            disabled=""
            name="quiz"
            type="hidden"
            value="[object Object]"
          />
          <p>
            Nom du quiz
          </p>
          <div
            aria-errormessage="quiz_error"
            aria-invalid="false"
          >
            <button
              class="c-CommonButton c-CommonButton_secondary"
              id="quiz_button"
              type="button"
            >
              Choisir le quiz
            </button>
          </div>
        </div>
      </div>
      <div
        class="c-QuizTipsTab__Group"
      >
        <div
          class="c-FormCheckbox"
        >
          <input
            class="c-FormCheckbox__Input "
            data-testid="form-checkbox"
            id="displayTips"
            name="displayTips"
            type="checkbox"
          />
          <label
            class="c-FormLabel"
            for="displayTips"
          >
            <div
              class="c-FormLabel__Label"
            >
              <span>
                Afficher des astuces 
              </span>
            </div>
          </label>
        </div>
        <div
          class="c-FormModalInput"
          data-testid="form-modal-input"
        >
          <label
            class="c-FormLabel"
            for="tips_button"
          >
            <div
              class="c-FormLabel__Label"
            >
              <span>
                Astuces à afficher 
              </span>
            </div>
          </label>
          <input
            class="c-FormModalInput__Input"
            disabled=""
            name="tips"
            type="hidden"
            value=""
          />
          <div
            aria-errormessage="tips_error"
            aria-invalid="false"
          >
            <button
              class="c-CommonButton c-CommonButton_secondary"
              id="tips_button"
              type="button"
            >
              Sélectionner les astuces à afficher
            </button>
          </div>
        </div>
      </div>
      <div
        class="c-QuizTipsTab__Buttons"
      >
        <button
          class="c-CommonButton c-CommonButton_primary"
          disabled=""
          type="submit"
        >
          Enregistrer les modifications
        </button>
        <button
          class="c-CommonButton c-CommonButton_secondary"
          disabled=""
          type="button"
        >
          Annuler les modifications
        </button>
      </div>
    </form>
  </div>
</div>
`;
